src/valkey-benchmark.c

```diff
@@ -1056,156 +1067,67 @@ static clusterNode **addClusterNode(clusterNode *node) {
         success = 0;
+        if (reply) fprintf(stderr, "%s\nCLUSTER SLOTS ERROR: %s\n", errmsg, reply->str);
         goto cleanup;
     }
-    reply = redisCommand(ctx, "CLUSTER NODES");
-    success = (reply != NULL);
-    if (!success) goto cleanup;
-    success = (reply->type != REDIS_REPLY_ERROR);
-    if (!success) {
-        if (config.hostsocket == NULL) {
-            fprintf(stderr, "Cluster node %s:%d replied with error:\n%s\n", config.conn_info.hostip,
-                    config.conn_info.hostport, reply->str);
-        } else {
-            fprintf(stderr, "Cluster node %s replied with error:\n%s\n", config.hostsocket, reply->str);
-        }
-        goto cleanup;
-    }
-    char *lines = reply->str, *p, *line;
-    while ((p = strstr(lines, "\n")) != NULL) {
-        *p = '\0';
-        line = lines;
-        lines = p + 1;
-        char *name = NULL, *addr = NULL, *flags = NULL, *primary_id = NULL;
-        int i = 0;
-        while ((p = strchr(line, ' ')) != NULL) {
-            *p = '\0';
-            char *token = line;
-            line = p + 1;
-            switch (i++) {
-            case 0: name = token; break;
-            case 1: addr = token; break;
-            case 2: flags = token; break;
-            case 3: primary_id = token; break;
-            }
-            if (i == 8) break; // Slots
-        }
-        if (!flags) {
-            fprintf(stderr, "Invalid CLUSTER NODES reply: missing flags.\n");
-            success = 0;
-            goto cleanup;
-        }
-        int myself = (strstr(flags, "myself") != NULL);
-        int is_replica = (strstr(flags, "slave") != NULL || (primary_id != NULL && primary_id[0] != '-'));
-        if (is_replica) continue;
-        if (addr == NULL) {
-            fprintf(stderr, "Invalid CLUSTER NODES reply: missing addr.\n");
-            success = 0;
-            goto cleanup;
-        }
-        clusterNode *node = NULL;
-        char *ip = NULL;
-        int port = 0;
-        char *paddr = strrchr(addr, ':');
-        if (paddr != NULL) {
-            *paddr = '\0';
-            ip = addr;
-            addr = paddr + 1;
-            /* If internal bus is specified, then just drop it. */
-            if ((paddr = strchr(addr, '@')) != NULL) *paddr = '\0';
-            port = atoi(addr);
-        }
-        if (myself) {
-            node = firstNode;
-            if (ip != NULL && strcmp(node->ip, ip) != 0) {
-                node->ip = sdsnew(ip);
-                node->port = port;
-            }
-        } else {
+    assert(reply->type == REDIS_REPLY_ARRAY);
+    for (i = 0; i < reply->elements; i++) {
+        redisReply *r = reply->element[i];
+        assert(r->type == REDIS_REPLY_ARRAY);
+        assert(r->elements >= 3);
+        int from = r->element[0]->integer;
+        int to = r->element[1]->integer;
+        for (j = 2; j < r->elements; j++) {
+            int is_primary = (j == 2);
+            if (config.read_from_replicas == is_primary) continue;
```

Comment:
Discussion is needed in this regard.
I've seen many clusters that operate with "Primary" handling only writes and "Replica" handling only reads. (Of course, this is despite the fact that Primary can read/write).
Therefore, there were three directions for implementation, and I adopted method 1, which is similar to the real world I've seen.
1/ Read from the replica option only reads from the replica, and if you need to read from the primary as well, run a separate process (current implementation)
2/ Read from the replica option reads from the entire cluster, including the primary (the way you described)
3/ Separate options depending on whether to include the primary

Changing to the method you told me is not difficult, please let me know your thoughts :)

Reply:
In addition, if we include the primary, we need to think about the handling write command.
e.g. `valkey-benchmark -h cluster_host -c 4 -t get,set -n 1000 --cluster --rfr`

I understand that I need to change the client to redirect to primary. It may be possible to simply divide reading into primary, but it seems likely that users want to be able to handle writing itself(valkey-benchmark). I think handling write commands should be implemented at other PR.

Reply:
Great. I'll implement the following tasks in relation to this discussion:
1/ Provide separate options for reading from replicas only and for reading from the entire cluster.

2/ Add CLI description about the option could not work properly when executing with write requests or commands.

And in the next PR/phase, I'll implement a way to distribute writes and reads well at the cluster level!

Btw, thank you for your advices and opinions ☺️